keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,774 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+At the beginning, `kube-proxy` was designed to handle the translation of Service objects to OS-level
+resources.  Implementations have been userland, then iptables, and now ipvs. With the growth of the
+Kubernetes project, more implementations came to life, for instance with eBPF, and often in relation
+to other goals (Calico to manage the network overlay, Cilium to manage app-level security, metallb
+to provide an external LB for bare-metal clusters, etc).
+
+Along this cambrian explosion of third-party software, the Service object itself received new
+concepts to improve the abstraction, for instance to express topology. This, third-party
+implementation are expected to update and become more complex over time, even if their core doesn't
+change (ie, the eBPF translation layer is not affected).
+
+This KEP is born from the conviction that more decoupling of the Service object and the actual
+implementations is required, by introducing an intermediate, node-level abstraction provider. This
+abstraction is expected to be the result of applying Kubernetes' `Service` semantics and business
+logic to a simpler, more stable API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
+- allow easier, more stable proxy implementations that don't need updates when `Service` business
```

Comment:
What is the scope of this? I think today kube-proxy implementations (iptables, ipvs, etc), CNIs (calico, cilium), service meshes, and ingress/gateways are all doing this to some degree. Do we expect all of them to get to this point, or only the first 1 or 2?

Reply:
I think at least for a first scope, kube-proxy will only provide information about services and endpoints, maybe network policy later, probably not ingress (I don't know yet about Gateway). So current implementations (userspace, iptables, ipvs, nft...) are in the target, and it should help calico, cilium and meshes to watch less kubernetes resources.

Reply:
The tricky thing is if it doesn't provide enough information, then it will either be completely unused or watch *more* resources. For example, if we precompute some new flattened API from Service and Endpoints, but we actually need some info from the endpoint that isn't in this flattened resource, we would need to still fetch the original endpoint (more watches) or just not use this new functionality at all.

The description above is a bit fuzzy since we don't have the details on what exactly the new information schema will be

Reply:
yes I'm not sure about everything but we should try to be quite minimalistic. The goals section is fuzzy but I have a live project with the definitions up-to-date here https://github.com/mcluseau/kube-proxy2/blob/master/pkg/api/localnetv1/services.proto ; TLDR: I'm quite generous on service information and minimalistic on endpoint info.

Reply:
So as one example, Istio (and others) need the ServiceAccount from an endpoint. This info is not present. As a result, we would still need to watch the full Endpoint, at which point it probably doesn't make sense to use this new api as well, since its redundant. SA is just one example, I am sure the list is quite long if we dig into it all.

I am not trying to say this KEP *must* support these use cases - but we should be very clear about whether it is intended to and whether it actually will.

Reply:
We don't have a (reliable) linking from endpoint to pod without targetRef. We also need labels from the pod.

Istio also happens to do a bunch of very Service-aware things like doing our own Service selection to determine the services a proxy is a part of when it connects to the control plane, and to handle VMs which we allow selecting by Service basically like pod. So a goal of Istio being able to replace all of its {Pod,Service,Endpoints} watches with this is probably not reasonable. Its plausible we could do just endpoints though.

Reply:
let's say we add `Endpoint{ EndpointMetadata Metadata }`, that Metadata would need:
- Namespace
- PodName
- PodLabels
- ServiceAccount

ServiceAccount seems to be the most specific part of it. I feel like any extended manager for services would need to put info on Pod with an admission controller (and also sidecars), if istio annotates the pod with its own `istio.security-account={pod.service-account}`, would having PodAnnotations in endpoint metadata be enough?

Reply:
i think the overall point is "we cant wrap the entire k8s api in the kpng data model" and thats a fair point.  That said, for proxies, we do wrap enough that the base network functionality works for service routing... i think thats good enough for a v1?